pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/SegmentDeletionManager.java

```diff
@@ -213,32 +210,54 @@ public void removeAgedDeletedSegments(int retentionInDays) {
         return;
       }
 
-      AgeFileFilter fileFilter = new AgeFileFilter(DateTime.now().minusDays(retentionInDays).toDate());
-      File[] directories = deletedDir.listFiles((FileFilter) DirectoryFileFilter.DIRECTORY);
-      // Check that the directory for deleted segments is empty
-      if (directories == null) {
-        LOGGER.warn("Deleted segment directory {} does not exist or it caused an I/O error.", deletedDir);
+      URI dataDirURI = ControllerConf.getUriFromPath(_dataDir);
+      URI deletedDirURI = ControllerConf.getUriFromPath(StringUtil.join(File.separator, _dataDir, DELETED_SEGMENTS));
+      PinotFS pinotFS = PinotFSFactory.create(dataDirURI.getScheme());
+
+      // Check that the directory for deleted segments exists
+      if (!pinotFS.isDirectory(deletedDirURI)) {
+        LOGGER.warn("Deleted segment directory {} does not exist or it is not directory.", deletedDirURI.toString());
+        return;
+      }
+
+    try {
+      String[] deletedDirFiles = pinotFS.listFiles(deletedDirURI, false);
+      if (deletedDirFiles == null) {
+        LOGGER.warn("Deleted segment directory {} does not exist.", deletedDirURI.toString());
         return;
       }
 
-      for (File currentDir : directories) {
+      for (String currentDir : deletedDirFiles) {
+        URI currentURI = ControllerConf.getUriFromPath(currentDir);
         // Get files that are aged
-        Collection<File> targetFiles = FileUtils.listFiles(currentDir, fileFilter, null);
+        String[] targetFiles = pinotFS.listFiles(currentURI, false);
+        List<URI> filesToDelete = new ArrayList<>();
+        for (String targetFile : targetFiles) {
+          URI targetURI = ControllerConf.getUriFromPath(targetFile);
+          Date dateToDelete = DateTime.now().minusDays(retentionInDays).toDate();
+          if (pinotFS.lastModified(targetURI) < dateToDelete.getTime()) {
```

Comment:
In the earlier implementation we used AgeFilter, so we had to call listFiles with the filter and then delete them and check if the dir is empty.
In your new implementation, you have the filter in your loop, so you should be able to:
(a) delete the files as you detect that they can be deleted.
(b) detect that the dir is empty and delete it afterwards (without doing another listFiles call).